include/pybind11/numpy.h

```diff
@@ -655,23 +655,18 @@ class array : public buffer {
     }
 
     /// Resize array to given shape
-    /// If refcheck is true, then resize will fail if > 1 references exists to this array
-    void resize(const std::vector<size_t> &new_shape, bool refcheck = false) {
+    /// If refcheck is true and more that one reference exist to this array
+    /// then resize will succeed only if it makes a reshape, i.e. original size doesn't change
+    void resize(const std::vector<size_t> &new_shape, bool refcheck = true) {
```

Comment:
I'm not a huge fan of taking concrete containers (`std::vector<size_t>`) like this--it's restrictive and something of a nuisance if you happen to have your shape in some other container.  Just to be clear: this isn't your fault at all; what you wrote here is perfectly consistent with the current API.  But I'm aiming to solve it with PR #788, which would allow you to change the signature to `void resize(ShapeContainer new_shape, bool refcheck = true)`, where the `ShapeContainer` is implicitly convertible from any sort of container, copy the contents into a `std::vector<Py_intptr_t>`.  That would also allow you to drop both the `reinterpret_cast` and `const_cast` from:
`reinterpret_cast<Py_intptr_t*>(const_cast<size_t*>(new_shape.data())), int(new_shape.size())`
replacing it with just:
`new_shape->data(), int(new_shape.size())`